pkg/compact/compact.go

```diff
@@ -470,6 +502,60 @@ func (cg *Group) Resolution() int64 {
 	return cg.resolution
 }
 
+// should return the results/metrics of the planning simulation
+type PlanSim interface {
+	ProgressCalculate(ctx context.Context, groups []*Group) error
+}
+
+type DefaultPlanSim struct {
+	planner Planner
+}
+
+func (ps *DefaultPlanSim) ProgressCalculate(ctx context.Context, groups []*Group) error {
+	iterations := 0
+	blocksToMerge := 0
+
+	// figure out hasPlan and noPlan
+	for {
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+
+			// parameter should be of type tsdb.BlockMeta.meta
+			plan, err := ps.planner.Plan(ctx, g.metasByMinTime)
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			iterations++
+
+			var toRemove map[ulid.ULID]bool
+			metas := make([]*tsdb.BlockMeta, 0, len(plan))
+			for _, p := range plan {
+				metas = append(metas, &p.BlockMeta)
+				toRemove[p.BlockMeta.ULID] = true
+			}
+			g.deleteFromGroup(toRemove)
+
+			blocksToMerge += len(plan)
+
+			// remove 'plan' blocks from 'original metadata' - so that the remaining blocks can now be planned ?
+			// not required to modify originalMeta now
+
+			newMeta := tsdb.CompactBlockMetas(ulid.MustNew(uint64(time.Now().Unix()), nil), metas...)
```

Comment:
One case we can check here:
We already removed planned metadata from the group. Now if there is no meta left in the group, then we don't need to append this new meta to the group, too.
Because even after the new meta is added, then the size is still 1 and will be ignored later. So no plan to do this case.